greymatter-io/terraform-greymatter-ecs
# Edge load balancer and mesh configs

closes #18 
closes #13 

This PR adds a module specifically for edge, with a load balancer for ingress.

When applied edge, control, control-api and control-api-sidecar come up. Using bootstrap mesh configs from startup, edge should be able to route directly to the control-api service - see [the backup file](https://github.com/greymatter-io/terraform-greymatter-ecs/blob/edge_lb_mesh/greymatter/modules/fabric/mesh/backup.json).  The idea is that after startup, the user runs `./mesh/apply.sh` (which will include more mesh configs as we build this out), which applies the correct mesh configs to route to control-api  through its sidecar, and deletes the direct connection.

Ready for review with service discovery in different network modes now working from [this PR](https://github.com/DecipherNow/gm-control/pull/404).

To test - follow the instructions from the readme for the full vpc setup, or build using your own vpc with:

```tf
module "greymatter" {
  source                       = "git::ssh://git@github.com/greymatter-io/terraform-greymatter-ecs//greymatter?ref=edge_lb_mesh"
  cluster_name                 = "gm-cluster"
  key_pair_name                = "ansible-provision"
  vpc_id                       = {}
  public_subnets               = {}
  private_subnets              = {}
  access_key_arn               = "{}"
  secret_access_key_arn        = "{}"
  aws_region                   = "us-east-1"
  dns_ns_name                  = "greymatter.dev"
  kms_ssm_arn                  = "{}"
  kms_secretsmanager_arn       = "{}"
  optimized_ami                = ""
  docker_gm_credentials        = { "username" : "{}", "password" : "{}" }
} 
```

Make sure that you copy the directory `gm` from this repo into wherever you are running terraform apply from.

**To test in buttermilk sky:**

1. Check out branch `ecs_module_test` and fill in docker_gm_credentials in `tf/resources/ecs_module_test.tf`. Note that I already copied directory `gm` into this branch for certs/mesh configs.

2. `terraform apply`.

3. Make sure you've awsumed bs-resources-admin and run `aws elbv2 describe-load-balancers --names edge | grep DNSName`. Copy the dns name and navigate to `dns_name:10808/services/control-api/latest` in a browser. You should see the control-api docs.

4. Use the same value for greymatter_api_host and configure the greymatter cli with:

```bash
export GREYMATTER_API_HOST={edgs_dns}:10808
export GREYMATTER_API_INSECURE=true
export GREYMATTER_API_PREFIX=/services/control-api/latest
export GREYMATTER_API_SSL=false
export GREYMATTER_CONSOLE_LEVEL=debug
```

Now run `./gm/mesh/apply.sh` to finish configuring the mesh.

5. You can check the logs for control-api-sidecar in cloudwatch to see the repeating update failures should go away. Go to `edge_dns:10808/services/control-api/latest/v1.0/cluster` to view the discovered instance for control-api sidecar or `greymatter get cluster edge-to-control-api-cluster`.

6. Then to take it down: `terraform destroy -target module.greymatter` and manually delete the autoscaling group ecs-autoscaling-group while it shuts down.
